drivers/adc/adc_cc13xx_cc26xx.c

```diff
@@ -0,0 +1,310 @@
+		       const struct adc_sequence *sequence,
+		       bool asynchronous,
+		       struct k_poll_signal *sig)
+{
+	struct adc_cc13xx_cc26xx_data *data = dev->data;
+	int rv;
+	size_t exp_size;
+
+	if (sequence->resolution != 12) {
+		LOG_ERR("Only 12 Resolution is supported, but %d got",
+			sequence->resolution);
+		return -EINVAL;
+	}
+
+	exp_size = sizeof(uint16_t);
+	if (sequence->options) {
+		exp_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < exp_size) {
+		LOG_ERR("Required buffer size is %u, but %u got",
+			exp_size, sequence->buffer_size);
+		return -ENOMEM;
+	}
+
+	data->buffer = sequence->buffer;
+
+	adc_context_lock(&data->ctx, asynchronous, sig);
+	adc_context_start_read(&data->ctx, sequence);
+	rv = adc_context_wait_for_completion(&data->ctx);
+	adc_context_release(&data->ctx, rv);
+	return rv;
+}
+
+static int adc_cc13xx_cc26xx_read(const struct device *dev,
+			   const struct adc_sequence *sequence)
+{
+	return cc13xx_cc26xx_read(dev, sequence, false, NULL);
+}
+
+#ifdef CONFIG_ADC_ASYNC
+static int adc_cc13xx_cc26xx_read_async(const struct device *dev,
+				 const struct adc_sequence *sequence,
+				 struct k_poll_signal *async)
+{
+	return cc13xx_cc26xx_read(dev, sequence, true, async);
+}
+#endif
+
+/**
+ * AUX_ADC_IRQ handler, called for either of these events:
+ * - conversion complete or DMA done (if used);
+ * - FIFO underflow or overflow;
+ */
+static void adc_cc13xx_cc26xx_isr(const struct device *dev)
+{
+	struct adc_cc13xx_cc26xx_data *data = dev->data;
+	/* get the statuses of ADC_DONE and ADC_IRQ events in order to clear them both */
+	uint32_t intStatus = (
+		HWREG(AUX_EVCTL_BASE + AUX_EVCTL_O_EVTOMCUFLAGS) &
+		(AUX_EVCTL_EVTOMCUFLAGS_AUX_ADC_IRQ | AUX_EVCTL_EVTOMCUFLAGS_AUX_ADC_DONE)
+	);
+	uint32_t fifoStatus;
+	uint32_t adcValue;
+
+	/* clear the AUXADC-related event flags */
+	HWREG(AUX_EVCTL_BASE + AUX_EVCTL_O_EVTOMCUFLAGSCLR) = intStatus;
+	/* check the ADC FIFO's status */
+	fifoStatus = AUXADCGetFifoStatus();
+	LOG_DBG("ISR flags 0x%08X fifo 0x%08X", intStatus, fifoStatus);
+	if ((fifoStatus & (AUX_ANAIF_ADCFIFOSTAT_OVERFLOW | AUX_ANAIF_ADCFIFOSTAT_UNDERFLOW))) {
+		AUXADCFlushFifo();
+	}
+	if ((fifoStatus & AUX_ANAIF_ADCFIFOSTAT_EMPTY_M)) {
+		/* no ADC values available */
+		return;
+	}
+	adcValue = AUXADCPopFifo();
+	*data->buffer++ = adcValue;
+	AUXADCDisable();
+
+	LOG_DBG("ADC val %d", adcValue);
+	adc_context_on_sampling_done(&data->ctx, dev);
+}
+
+static const struct adc_driver_api cc13xx_cc26xx_driver_api = {
+	.channel_setup = adc_cc13xx_cc26xx_channel_setup,
+	.read = adc_cc13xx_cc26xx_read,
+#ifdef CONFIG_ADC_ASYNC
+	.read_async = adc_cc13xx_cc26xx_read_async,
+#endif
+	.ref_internal = 4300, /* fixed reference: 4.3V */
+};
+
+#define CC13XX_CC26XX_ADC_IRQ_CONNECT(index)					 \
+	IRQ_CONNECT(DT_INST_IRQN(index), DT_INST_IRQ(index, priority),		 \
+		adc_cc13xx_cc26xx_isr, DEVICE_DT_INST_GET(index), 0);		 \
+	irq_enable(DT_INST_IRQN(index));					 \
```

Comment:
Do you actually need a macro for this? Couldn't that content be placed directly in the `adc_cc13xx_cc26xx_cfg_func_##index` function body?